contracts/nft/nft_collection_base.ral

```diff
@@ -0,0 +1,28 @@
+import "std/nft_collection_interface"
+
+Abstract Contract NFTCollectionBase(
+    collectionUri: ByteVec,
+    mut totalSupply: U256
+) implements INFTCollection {
+    enum ErrorCodes {
+        IncorrectTokenIndex = 0
+        NFTNotFound = 1
+    }
+
+    pub fn getCollectionUri() -> ByteVec {
+        return collectionUri
+    }
+
+    pub fn totalSupply() -> U256 {
+        return totalSupply
+    }
+
+    pub fn nftByIndex(index: U256) -> INFT {
+        checkCaller!(index < totalSupply(), ErrorCodes.IncorrectTokenIndex)
```

Comment:
[Nit]: We can use the `totalSupply` to avoid a function call. And seems like we should use `assert!` instead of `checkCaller!` here.

Reply:
Yes! I can remove the checkCaller now after the signature change you made recently to the full node.